### DISCUSSION TITLE
NonNull input field with default value throws exception if not set

### DISCUSSION THREAD

**Discussion by mingh2** [graphql-java/graphql-java#2172] [Q&A] [answered]
When a NonNull input field is not provided, should it be coerced to default value if present?

However, with the following (simplified) situation, I got a NonNullableValueCoercedAsNullException
```
input ListSnacksInput { 
  kitchen: [Kitchen!]! = [] 
  startPageToken: String 
} 
type Query {
  listSnacks(input: ListSnacksInput!) : [Snacks!]!
}


query ($input: ListSnacksInput!){
  listSnacks(input: $input) {
     ...
  }
}
variables:
{
  input: {startPageToken: "dummyToken"}
}
```

Here's the source [code](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/ValuesResolver.java#L199) that checks if the value is null and resolve the value to default value if possible. But it is called after [#coerceValue](https://github.com/graphql-java/graphql-java/blob/59cd870263031113a5ef3477ea862b24a0d1b8f5/src/main/java/graphql/execution/ValuesResolver.java#L142), which will throw exception first if the value is null.

**Comment by aoudiamoncef** [comment]
It could be related to #2084

**Reply by mingh2** [reply]
Thanks for pointing this out! Do you suggest this could be fixed already?

I view this two a little bit different,
#2084 seems to indicate that the default value for a variable **object** is not respected by the query. 

The pr fixes the issue because in [here](https://github.com/graphql-java/graphql-java/blob/4213a3034aa18edc36207753ea274c3824593e1b/src/main/java/graphql/execution/ValuesResolver.java#L62)
1. it first checks if the default is provided.
2. if not, throws exception if value should be non null and value resolves to null.

Our case is more like default value for a **field** is not respected.
The order of the coercing for a field is different from the object
1. throws exception if value should be non null and value resolves to null
2. assign a default value if the value resolves to null (never reaches)

Stacktrace:
```
coerceValue:142, ValuesResolver (graphql.execution) [4]
coerceValueForInputObjectType:195, ValuesResolver (graphql.execution)
coerceValue:159, ValuesResolver (graphql.execution) [3]
coerceValueForInputObjectType:195, ValuesResolver (graphql.execution)
coerceValue:159, ValuesResolver (graphql.execution) [2]
coerceValue:140, ValuesResolver (graphql.execution) [1]
coerceVariableValues:70, ValuesResolver (graphql.execution)
```


Also wanted to point out that, it works for query
```
query {
  listSnacks(input: {startPageToken: "dummyToken"}) {
     ...
  }
}
```

I might be wrong here but it seems to suggest inconsistent behavior when resolving value for input if specified directly in query vs in variables. Let me know if you have more context on this.

**Reply by bbakerman** [reply]
I suspect that this defaulting is not taking into account that "complex" structure of the input map.  

I feel like this is a bug - let me see if I can make a test case for this to see the behavior in action

**Comment by bbakerman** [comment, marked as answer]
See #2180 for a fix for this bug

**Reply by mingh2** [reply]
Thanks for the fix!
